### PR TITLE
highgui: wayland: fix to pass highgui test

### LINKED ISSUE #25550: highgui: wayland: opencv_test_highgui fails
### System Information

OpenCV version: 4.x ( bc95f27e56d89e486c25e76db0573b610bee6496 )
Operating System / Platform: Ubuntu 24.04
Compiler & compiler version: GCC 13.2

```
--   GUI:                           Wayland
--     Wayland:                     (Experimental) YES
--       Wayland Client:            YES (ver 1.22.0)
--       Wayland Cursor:            YES (ver 1.22.0)
--       Wayland Protocols:         YES (ver 1.34)
--       Xkbcommon:                 YES (ver 1.6.0)
--       Wayland EGL(Option):       YES (ver 18.1.0)
```

### Detailed description

opencv_test_highgui with wayland fails.

- (A) it requires to support CV_8S/16S/32F/64F support (When debug build, there are some exceptions)
- (B) it requires to initlize slider position with value in user pointer.
- (C) it requires to update slider value and call update user function, but now if window condition is not ready, it will not.
- (D) it requires to call getWindowImageRect(), but wayland backend is not supported by protocol limitation.
- (E) it requires to get "1" if createTrackbar() is succeeded, but wayland backend returns 0.
- 
```plaintext

kmtr@kmtr-VMware-Virtual-Platform:~/work/build4-main$ ./bin/opencv_test_highgui
CTEST_FULL_OUTPUT
OpenCV version: 4.9.0-dev
OpenCV VCS version: 4.9.0-423-gbc95f27e56
Build type: Debug
Compiler: /usr/bin/c++  (ver 13.2.0)
[ INFO:0@0.004] global registry_parallel.impl.hpp:96 ParallelBackendRegistry core(parallel): Enabled backends(3, sorted by priority): ONETBB(1000); TBB(990); OPENMP(980)
Parallel framework: pthreads (nthreads=8)
CPU features: SSE SSE2 SSE3 *SSE4.1 *SSE4.2 *FP16 *AVX *AVX2 *AVX512-SKX?
Intel(R) IPP version: ippIP AVX2 (l9) 2021.11.0 (-) Feb  1 2024
Intel(R) IPP features code: 0x8000
[ INFO:0@0.005] global ocl.cpp:1185 haveOpenCL Initialize OpenCL runtime...
[ INFO:0@0.005] global ocl.cpp:1191 haveOpenCL OpenCL: found 0 platforms
OpenCL is disabled
TEST: Skip tests with tags: 'mem_6gb', 'verylong', 'debug_verylong'
[==========] Running 4 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 4 tests from Highgui_GUI
[ RUN      ] Highgui_GUI.regression
[ INFO:0@0.006] global registry.impl.hpp:114 UIBackendRegistry UI: Enabled backends(3, sorted by priority): GTK(1000); GTK3(990); GTK2(980) + BUILTIN(Wayland)
[*] DEBUG: handle_toplevel_configure: maximized=0 fullscreen=0 resizing=0 focused=0 size=[0 x 0]
[*] DEBUG: handle_toplevel_configure: maximized=0 fullscreen=0 resizing=0 focused=1 size=[800 x 624]
/home/kmtr/work/opencv4/modules/highgui/test/test_gui.cpp:53: Failure //////////////// (D)
Expected equality of these values:
  rc.size()
    Which is: -1x-1
  img.size()
    Which is: 800x600
Google Test trace:
/home/kmtr/work/opencv4/modules/highgui/test/test_gui.cpp:80: 0
Exception message: OpenCV(4.9.0-dev) /home/kmtr/work/opencv4/modules/core/include/opencv2/core/mat.inl.hpp:910: error: (-215:Assertion failed) (unsigned)(i1 * DataType<_Tp>::channels) < (unsigned)(size.p[1] * channels()) in function 'at' 

/home/kmtr/work/opencv4/modules/highgui/test/test_gui.cpp:49: Failure /////////////////(A)
Expected: imshow(nm, img) doesn't throw an exception.
  Actual: it throws.
Google Test trace:
/home/kmtr/work/opencv4/modules/highgui/test/test_gui.cpp:80: 1
:
:
:
[  FAILED  ] Highgui_GUI.regression (8729 ms)
[ RUN      ] Highgui_GUI.trackbar_unsafe
[ WARN:0@8.778] global window.cpp:701 createTrackbar UI/Trackbar(trackbar@trackbar_test_window): Using 'value' pointer is unsafe and deprecated. Use NULL as value pointer. To fetch trackbar value setup callback.
/home/kmtr/work/opencv4/modules/highgui/test/test_gui.cpp:154: Failure //////////////// (E)
Expected equality of these values:
  (int)1
    Which is: 1
  createTrackbar(trackbar_name, window_name, &value, 100, Foo, &callback_count)
    Which is: 0
/home/kmtr/work/opencv4/modules/highgui/test/test_gui.cpp:155: Failure ////////////////////////// (B)
Expected equality of these values:
  value
    Which is: 50
  getTrackbarPos(trackbar_name, window_name)
    Which is: 0
/home/kmtr/work/opencv4/modules/highgui/test/test_gui.cpp:160: Failure ////////////////////// (C)
Expected equality of these values:
  callback_count_base + 1
    Which is: 1
  callback_count
    Which is: 0
/home/kmtr/work/opencv4/modules/highgui/test/test_gui.cpp:161: Failure /////////////////////////////// (C)
Expected equality of these values:
  90
  value
    Which is: 50
[  FAILED  ] Highgui_GUI.trackbar_unsafe (44 ms)
:
:
[ RUN      ] Highgui_GUI.currentUIFramework
UI framework: "WAYLAND"
[       OK ] Highgui_GUI.currentUIFramework (0 ms)
[----------] 4 tests from Highgui_GUI (8793 ms total)

[----------] Global test environment tear-down
[==========] 4 tests from 1 test case ran. (8793 ms total)
[  PASSED  ] 1 test.
[  FAILED  ] 3 tests, listed below:
[  FAILED  ] Highgui_GUI.regression
[  FAILED  ] Highgui_GUI.trackbar_unsafe
[  FAILED  ] Highgui_GUI.trackbar

 3 FAILED TESTS
kmtr@kmtr-VMware-Virtual-Platform:~/work/build4-main$
```

### Steps to reproduce

build opencv and run ` ./bin/opencv_test_highgui`

### Issue submission checklist

- [X] I report the issue, it's not a question
- [X] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [X] I updated to the latest OpenCV version and the issue is still there
- [X] There is reproducer code and related data files (videos, images, onnx, etc)

### PR DESCRIPTION
**Pull request by Kumataro**
Close #25550 

- optimize Mat to XRGB8888 conversion with OpenCV functions
  - extend to support CV_8S/16U/16S/32F/64F
  - extend to support 1/4 channels
- fix to update value timing
 - initilize slider_ value if value is not nullptr.
 - Update user-ptr value and call on_change() function if cv_wl_trackbar::draw() is not called.
- Update usage of WAYLAND/XDG macro to avoid reference undefined macro.
- Update documents

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by Kumataro** [comment]
This patch contains performance tuning.

Result
-------
I compare count of Instruction references. 

| Environment            | Ir of write_mat_to_xrgb8888()  |
|:----------------------:|-----------:|
| OpenCV 4.9(Before) | 5,911,406 |
| Without SIMD          | 2,453,868 |
| SSE3 + SIMD128      |  1.150.789 |
| SSE4.1 + SIMD128 |  325,695 |
| AVX2 + SIMD256 | 260,544|

The differences between SSE3 and SSE4.1 comes from intristic implementation.

https://github.com/opencv/opencv/blob/dad8af6b17f8e60d7b95a1203a1b4d22f56574cf/modules/core/include/opencv2/core/hal/intrin_sse.hpp#L2422-L2453

Test 
-----

Source code is here.

```cpp
// g++ main.cpp -o a.out -I /usr/local/include/opencv4 -lopencv_core -lopencv_highgui -lopencv_imgcodecs
#include <opencv2/core.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/imgcodecs.hpp>
#include <iostream>
#include <string>

int main(int argc, char *argv[])
{
  std::cout << "cv::currentUIFramework() returns " << cv::currentUIFramework() << std::endl;

  cv::Mat src;
  src = cv::imread("opencv-logo.png");

  cv::namedWindow("src");
  cv::imshow("src", src);
  (void)cv::waitKey(1000);

  return 0;
}
```

Command is here.

```bash
valgrind --tool=callgrind ./a.out
callgrind_annotate callgrind.out.[PID] | grep 8888 | head -1
```

File: modules/highgui/src/write_mat_to_xrgb8888.simd.hpp [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+// of this distribution and at http://opencv.org/license.html.
+
+#include "precomp.hpp"
+#include "opencv2/core/hal/intrin.hpp"
+
+
+namespace cv {
+CV_CPU_OPTIMIZATION_NAMESPACE_BEGIN
+
+void write_mat_to_xrgb8888(cv::Mat const &img, void *data);
```
Comment (opencv-alalek) [inline modules/highgui/src/write_mat_to_xrgb8888.simd.hpp:12]: Please avoid using Mat in optimized dispatched code. Use void*, step, size instead.

File: modules/highgui/src/write_mat_to_xrgb8888.simd.hpp [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+
+#ifndef CV_CPU_OPTIMIZATION_DECLARATIONS_ONLY
+
+void write_mat_to_xrgb8888(cv::Mat const &img, void *data) {
+    CV_CheckTrue(data != nullptr, "data must not be nullptr.");
+    CV_CheckType(img.type(), (img.type() == CV_8UC1) || (img.type() == CV_8UC3) || (img.type() == CV_8UC4),
```
Comment (opencv-alalek) [inline modules/highgui/src/write_mat_to_xrgb8888.simd.hpp:18]: Checks should be done outside of optimized functions.

File: modules/highgui/src/write_mat_to_xrgb8888.simd.hpp (hunk 1 above)
Comment (Kumataro) [reply modules/highgui/src/write_mat_to_xrgb8888.simd.hpp]: Thank you for your review, I fixed it.

```
Wayland Backend
|
write_mat_to_xrgb8888(const cv::Mat& img, void *data)
  + cv::opt_SSE4_1::write_g8_to_xrgb8888(uint8_t* src, uint8_t* dst, int img_col);
  + cv::opt_SSE4_1::write_bgr888_to_xrgb8888(uint8_t* src, uint8_t* dst, int img_col);
  + cv::opt_SSE4_1::write_bgra8888_to_xrgb8888(uint8_t* src, uint8_t* dst, int img_col);
  + cv::opt_AVX2::write_g8_to_xrgb8888(uint8_t* src, uint8_t* dst, int img_col);
  + cv::opt_AVX2::write_bgr888_to_xrgb8888(uint8_t* src, uint8_t* dst, int img_col);
  + cv::opt_AVX2::write_bgra8888_to_xrgb8888(uint8_t* src, uint8_t* dst, int img_col);
```

File: modules/highgui/src/write_mat_to_xrgb8888.simd.hpp (hunk 2 above)
Comment (Kumataro) [reply modules/highgui/src/write_mat_to_xrgb8888.simd.hpp]: Thank you for your comment, I removed redandancy checks.

**Review by asmorkalov — APPROVED** [review]
:+1: Tested manually with Kubuntu 24.04 and Wayland session.
Proposed todo: Add RISC-V RVV and other scalable vector intrinsics support. Need to use CV_SIMD_SCALABLE macro and run-time value step in loops.

**Comment by Kumataro** [comment]
> Proposed todo: Add RISC-V RVV and other scalable vector intrinsics support. Need to use CV_SIMD_SCALABLE macro and run-time value step in loops.

Thank you for your proposal ! I'll try it this weekend.
I have no ARM SVE and RISC-V Vector Extension environment, but it looks like to able to test with AVX environment.

I think current implementation will be refactoring similar to split function.
https://github.com/opencv/opencv/blob/4.x/modules/core/src/split.simd.hpp

For example(this is only my imagination ).
```cpp
template<typename T, typename VecT> static void
vecwrite_T_to_xrgb8888( const T* src, T* dst, int len, int scn )
{
    const int VECSZ = VTraits<VecT>::vlanes();
    const int dcn = 4; // XRGB
:
:
    else if( scn == 3 )
    {
        for( i = 0; i < len; i += VECSZ )
        {
            if( i > len - VECSZ )
            {
                i = len - VECSZ;
                mode = hal::STORE_UNALIGNED;
            }
            VecT b,g,r;
            v_load_deinterleave(src + i*scn, b, g, r);
            v_store_interleave (dst + i*dcn, b, g, r, r, mode);
            if( i < i0 )
            {
                i = i0 - VECSZ;
                mode = hal::STORE_ALIGNED_NOCACHE;
            }
        }
    }
```

**Comment by Kumataro** [comment]
I update code to support CV_SIMD_SCALABLE and tested with VMWare(AVX2) and Raspi4(NEON) with ubuntu24.04. opencv_test_highgui is passed and it called vector implementation.

**Comment by Kumataro** [comment]
I see cvtColor implementation, and I have second idea to use cvtColor(cv::BGR2BGRA) or cvtColor(cv::GRAY2BGRA) instead of this SIMD implementation. I'l try it.

We can get many performance improvemet, which are provided from OpenCL, IPP, multithread, via cvtColor().  And furthermore, the maintainability of the code is also improved.

File: modules/highgui/test/test_gui.cpp [hunk 3]
```diff
@@ -48,9 +48,11 @@ inline void verify_size(const std::string &nm, const cv::Mat &img)
 {
     EXPECT_NO_THROW(imshow(nm, img));
     EXPECT_EQ(-1, waitKey(200));
+#ifndef HAVE_WAYLAND // Wayland backend is not supported getWindowImageRect().
     Rect rc;
     EXPECT_NO_THROW(rc = getWindowImageRect(nm));
     EXPECT_EQ(rc.size(), img.size());
+#endif // HAVE_WAYLAND
```
Comment (asmorkalov) [inline modules/highgui/test/test_gui.cpp:55]: I propose to use runtime check with `currentUIFramework` to support build with UI plugins. Also please add reference to ticket to the check comment.

File: modules/highgui/include/opencv2/highgui.hpp [hunk 4]
```diff
@@ -394,6 +399,8 @@ CV_EXPORTS_W void resizeWindow(const String& winname, const cv::Size& size);
 @param winname Name of the window.
 @param x The new x-coordinate of the window.
 @param y The new y-coordinate of the window.
+
+@note [__Wayland Backend Only__] This function is not supported by the wayland protocol limitation.
```
Comment (asmorkalov) [inline modules/highgui/include/opencv2/highgui.hpp:403]: Wayland

File: modules/highgui/src/window_wayland.cpp [hunk 5]
```diff
@@ -84,23 +84,87 @@ static int xkb_keysym_to_ascii(xkb_keysym_t keysym) {
...
+        uint8_t* dst = (uint8_t*)data;
+        for(int y = 0; y < img_rows; y++, dst+=img_cols * 4)
+        {
+            const uint8_t* src = img.ptr(y);
+            memcpy(dst, src, img_cols * 4);
         }
     }
 }
 
+
```
Comment (asmorkalov) [inline modules/highgui/src/window_wayland.cpp:167]: Redundant.

File: modules/highgui/src/window_wayland.cpp [hunk 6]
```diff
@@ -84,23 +84,87 @@ static int xkb_keysym_to_ascii(xkb_keysym_t keysym) {
...
+        const bool isContinuous = img.isContinuous();
+        const int img_cols = (!isContinuous) ? (img.cols) : (img.cols * img.rows);
+        const int img_rows = (!isContinuous) ? (img.rows) : (1);
+
+        uint8_t* dst = (uint8_t*)data;
+        for(int y = 0; y < img_rows; y++, dst+=img_cols * 4)
+        {
+            const uint8_t* src = img.ptr(y);
+            memcpy(dst, src, img_cols * 4);
         }
```
Comment (asmorkalov) [inline modules/highgui/src/window_wayland.cpp:163]: Why not just `cv::Mat::copyTo`?

File: modules/highgui/test/test_gui.cpp (hunk 3 above)
Comment (Kumataro) [reply modules/highgui/test/test_gui.cpp]: Thank you for your review, I fixed it.

File: modules/highgui/include/opencv2/highgui.hpp (hunk 4 above)
Comment (Kumataro) [reply modules/highgui/include/opencv2/highgui.hpp]: Yes, I agree with you. I fixed it.

File: modules/highgui/src/window_wayland.cpp (hunk 5 above)
Comment (Kumataro) [reply modules/highgui/src/window_wayland.cpp]: OK, I removed it.

File: modules/highgui/src/window_wayland.cpp (hunk 6 above)
Comment (Kumataro) [reply modules/highgui/src/window_wayland.cpp]: `dst` is mat which memory had been pre-allocated.
So I was thinking with more safety.
But there are no problem working with copyTo().
Thank you for your comment!!

**Review by asmorkalov — APPROVED** [review]
:+1:

**Pull request merged**